File: youtube/lat4-iris/try2.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(corrplot)
library(caTools)
library(dplyr)
library(fastDummies)
library(zoo)
library(pROC)
library(ROCR)
library(leaps)
library(car)
library(tidyr)
library(plyr)
library(lubridate)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees


df <- read.csv("ML/youtube/lat4-iris/iris_data.csv", header = FALSE, col.names = c("Sepal_Length", "Sepal_Width", "Petal_Length", "Petal_Width", "Species"))
print(head(df))
print(str(df))
print(summary(df))
print(colSums(is.na(df)))


p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Sepal_Length), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Sepal_Width), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Petal_Length), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Petal_Width), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # # Menggabungkan plot menggunakan patchwork
# plot <- p1+p2+p3+p4
# # plot <- (p1 | p2) / (p3 | p4)
# print(plot)



# q1 <- quantile(df$Sepal_Length, 0.25)
# q3 <- quantile(df$Sepal_Length, 0.75)
# iqr <- q3 - q1
# # Mengaplikasikan aturan IQR untuk menyaring data
# df <- subset(df, Sepal_Length >= q1 - 1.5 * iqr & Sepal_Length <= q3 + 1.5 * iqr)



# Buat plot boxplot untuk setiap variabel
boxplot1 <- ggplot(df, aes(x = "", y = `Sepal_Length`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot2 <- ggplot(df, aes(x = "", y = `Sepal_Width`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot3 <- ggplot(df, aes(x = "", y = `Petal_Length`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot4 <- ggplot(df, aes(x = "", y = `Petal_Width`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
# Menggabungkan plot menggunakan grid.arrange
plot <- grid.arrange(boxplot1, boxplot2, boxplot3, boxplot4, ncol = 2)
# plot <- boxplot1 + boxplot2+boxplot3+boxplot4
print(plot)

cat("\n\n\n\n")
print(unique(df$Species))
# df$Species <- case_when(
#     df$Species == 'Iris-setosa' ~ 0,
#     df$Species == 'Iris-versicolor' ~ 1,
#     df$Species == 'Iris-virginica' ~ 2
# )
df$Species <- ifelse(df$Species == "Iris-setosa", 0, ifelse(df$Species == "Iris-versicolor", 1, 2))
print(head(df))

numvars <- c("Sepal_Length", "Sepal_Width", "Petal_Length", "Petal_Width")
scaler <- preProcess(df[numvars], method = "range")
df[numvars] <- predict(scaler, df[numvars])
print(head(df))


## Set random seed
set.seed(42)
## Create train and test indices
split <- sample.split(df, SplitRatio = 0.7)

## Create train and test dataframes
# df_train <- df[split,]
# df_test <- df[!split,]
df_train <- subset(df, split == TRUE)
df_test <- subset(df, split == FALSE)

## Print first few rows of train and test dataframes
print(head(df_train))
print(head(df_test))
print(dim(df))
print(dim(df_train))
print(dim(df_test))


corr <- cor(df_train)
print(corr)
corplot <- corrplot(corr, method = 'color')
print(corplot)


x_train <- subset(df_train, select = -Species)
y_train <- as.factor(df_train$Species)
print(head(x_train))
print(head(y_train))


ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="nnet", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="rpart", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="nb", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')


# Membuat kontrol RFE
# control <- rfeControl(functions = leapsFuncs, method = "cv", number =10)
control <- rfeControl(functions = rfFuncs, method = "cv", number =10)
# Menjalankan RFE
rfe_model <- rfe(x = x_train, y = y_train, sizes = c(3), rfeControl = control)
# Mendapatkan hasil support dan ranking
support <- predictors(rfe_model)
ranking <- varImp(rfe_model, scale = FALSE)

print(support)
print(ranking)

# Membuat dataframe hasil
hasil <- data.frame(
    Features = rownames(ranking),
    Support = rownames(ranking) %in% support,
    Rank = ranking$Overall
)

print(hasil)

support <- hasil$Features[hasil$Support == TRUE]
not_support <- hasil$Features[hasil$Support == FALSE]
print(support)
print(not_support)

x_train_rfe <- x_train[support]
print(head(x_train_rfe))

ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train_rfe, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')


# Menghitung VIF
# Gabungkan x_train_rfe dan y_train menjadi satu data frame
data_train <- cbind(x_train_rfe, response = y_train)
# Membuat formula model linier dengan semua fitur dalam x_train_rfe
formula <- as.formula(paste("response ~", paste(names(x_train_rfe), collapse = " + ")))
# Membuat model linier menggunakan formula tersebut
model <- glm(formula, data = data_train, family = binomial)
# Menghitung VIF
vif_values <- vif(model)
print(vif_values)

# Membuat data frame dengan hasil VIF
vif_df <- data.frame(
    Features = names(vif_values),
    VIF = vif_values
)
vif_df <- vif_df[order(-vif_df$VIF), ]

print(vif_df)


y_pred <- predict(modelrf, x_train_rfe)
df_pred <- data.frame(
    y_train = y_train,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_train != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_train)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_train)
print(cm)


x_test <- dplyr::select(df_test, -Species)
x_test_rfe <- x_test[support]
y_test <- as.factor(df_test$Species)


y_pred <- predict(modelrf, x_test_rfe)
df_pred <- data.frame(
    y_test = y_test,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_test != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_test)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_test)
print(cm)

File: youtube/lat4-iris/try2.py
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns 
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_selection import RFE
import statsmodels.api as sm 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor

df = pd.read_csv('ML/youtube/lat4-iris/iris_data.csv', header = None, names = ['Sepal Length (cm)', 'Sepal Width (cm)',  'Petal Length (cm)',  'Petal Width (cm)',  'Species'])
print(df.head())
print(df.info())
print(df.describe())
print(df.isnull().sum())

# fig, ax = plt.subplots(2,2, figsize=(10,5))
# plt1 = sns.boxplot(df['Sepal Length (cm)'], ax = ax[0,0])
# plt2 = sns.boxplot(df['Sepal Width (cm)'], ax = ax[0,1])
# plt3 = sns.boxplot(df['Petal Length (cm)'], ax = ax[1,0])
# plt4 = sns.boxplot(df['Petal Width (cm)'], ax = ax[1,1])
# plt.tight_layout()
# plt.show()

# fig, ax = plt.subplots(2, 3, figsize=(15, 10))
# sns.scatterplot(x=df.index, y=df['Sepal Length (cm)'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['Sepal Width (cm)'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=df.index, y=df['Petal Length (cm)'], ax=ax[1, 0], color='orange')
# sns.scatterplot(x=df.index, y=df['Petal Width (cm)'], ax=ax[1, 1], color='red')
# plt.tight_layout()
# plt.show()

print(df['Species'].unique())
df['Species'] = df['Species'].map({'Iris-setosa': 0, 'Iris-versicolor':1, 'Iris-virginica':2})
print(df.head())

numvars = ['Sepal Length (cm)', 'Sepal Width (cm)',  'Petal Length (cm)',  'Petal Width (cm)']
scaler = MinMaxScaler()
df[numvars] = scaler.fit_transform(df[numvars])

df_train, df_test = train_test_split(df, train_size = 0.7, test_size=0.3, random_state=100)
print(df_train.head())
print(df_test.head())

corr = df_train.corr()
print(corr)
# plt.figure(figsize=(16, 10))
# sns.heatmap(corr, annot = True, cmap="YlGnBu")
# plt.show()

x_train = df_train
y_train = df_train.pop('Species')

k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelNB  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelNB, x_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelNB.fit(x_train, y_train)

# sebenarnya tidak dibutuhkan karena jumlah kolomnya sendiri hanya 4
rfe = RFE(estimator=modelNB, n_features_to_select=3)
rfe = rfe.fit(x_train, y_train)
print(rfe.support_)
print(rfe.ranking_)

hasil = pd.DataFrame({
    'features': x_train.columns,
    'support': rfe.support_,
    'rank': rfe.ranking_,
})
print(hasil)

support = x_train.columns[rfe.support_]
not_support = x_train.columns[~rfe.support_]
print(support)
print(not_support)

x_train_rfe = x_train[support]
# x_train_rfe = sm.add_constant(x_train_rfe)
print(x_train_rfe.head())
print(x_train_rfe.shape)


k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelNB  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelNB, x_train_rfe, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelNB.fit(x_train_rfe, y_train)

vif = pd.DataFrame()
x = x_train_rfe
print(x.shape)

vif['features'] = x.columns
vif['VIF'] = [variance_inflation_factor(x.values, i) for i in range(x.shape[1])]
vif = vif.sort_values('VIF', ascending=False)
print(vif)

vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values('VIF', ascending=False)
print(vif)


y_pred = modelNB.predict(x_train_rfe)

df = pd.DataFrame({
    'y_train': y_train,
    'y_pred': y_pred,
})
print(df.head())

akurasi = accuracy_score(y_train, y_pred)
print(akurasi)
cm = confusion_matrix(y_train, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

x_test = df_test
y_test = df_test.pop('Species')

x_test_rfe = x_test[support]
# x_test_rfe = sm.add_constant(x_test_rfe)
y_pred = modelNB.predict(x_test_rfe)

df = pd.DataFrame({
    'y_test': y_test,
    'y_pred': y_pred,
})
print(df.head())

akurasi = accuracy_score(y_test, y_pred)
print(akurasi)
cm = confusion_matrix(y_test, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()